### ISSUE TITLE
Compatibility with Sonar Java Analyzer 4.0

### ISSUE THREAD

**Issue by amandel** [SonarQubeCommunity/sonar-findbugs#29]
I'm not sure if this is also monitored here. There are some breaking changes in the to be released Java Analyzer 4.0 (https://groups.google.com/d/msg/sonarqube/hfOPZtakoeA/zqrz1640BgAJ)

**Comment by m-g-sonar** [comment]
Thanks for creating the tissue @amandel, I was about to do it. 

To summarize the situation, version 4.0 of the Java plugin is now up-to-date with the API of the new SonarQube LTS (5.6). In order to align our code, we had to remove usage of some deprecated API from SQ, but it also introduced breaking changes.

The change of the Java plugin API which is currently causing failure of the analysis of the Findbugs plugin is related to the `JavaResourceLocator`. See line [Line #40](https://github.com/SonarSource/sonar-java/blob/5bb12a036d6001ed0e5853a39be21929f29a31ad/java-frontend/src/main/java/org/sonar/plugins/java/api/JavaResourceLocator.java#L40). 
Prior to the change, the method `findResourceByClassName()` was returning a `org.sonar.api.resources.Resource`. Now it is returning an `org.sonar.api.batch.fs.InputFile`.

Trace of the analysis failure: 
```
[ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.1-build231:sonar (default-cli) on project HeadShot-java8: Execution default-cli of goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.1-build231:sonar failed: An API incompatibility was encountered while executing org.sonarsource.scanner.maven:sonar-maven-plugin:3.1-build231:sonar: java.lang.NoSuchMethodError: org.sonar.plugins.java.api.JavaResourceLocator.findResourceByClassName(Ljava/lang/String;)Lorg/sonar/api/resources/Resource;

...

Caused by: java.lang.NoSuchMethodError: org.sonar.plugins.java.api.JavaResourceLocator.findResourceByClassName(Ljava/lang/String;)Lorg/sonar/api/resources/Resource;
	at org.sonar.plugins.findbugs.FindbugsSensor.analyse(FindbugsSensor.java:108)
	at org.sonar.batch.phases.SensorsExecutor.executeSensor(SensorsExecutor.java:58)
	at org.sonar.batch.phases.SensorsExecutor.execute(SensorsExecutor.java:50)
	at org.sonar.batch.phases.AbstractPhaseExecutor.execute(AbstractPhaseExecutor.java:83)
	at org.sonar.batch.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:192)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
	at org.sonar.batch.scan.ProjectScanContainer.scan(ProjectScanContainer.java:241)
	at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:236)
	at org.sonar.batch.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:226)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
	at org.sonar.batch.task.ScanTask.execute(ScanTask.java:47)
	at org.sonar.batch.task.TaskContainer.doAfterStart(TaskContainer.java:86)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
	at org.sonar.batch.bootstrap.GlobalContainer.executeTask(GlobalContainer.java:106)
	at org.sonar.batch.bootstrapper.Batch.executeTask(Batch.java:119)
	at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:62)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
	at com.sun.proxy.$Proxy21.execute(Unknown Source)
	at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:240)
	at org.sonarsource.scanner.api.EmbeddedScanner.runAnalysis(EmbeddedScanner.java:151)
	at org.sonarsource.scanner.maven.bootstrap.ScannerBootstrapper.execute(ScannerBootstrapper.java:78)
	at org.sonarsource.scanner.maven.SonarQubeMojo.execute(SonarQubeMojo.java:112)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	... 20 more
```

As you already updated your API to 5.6 (and in particular the `FindbugsSensor`), it seems that the root of the failure has been already removed. I can only spot some usage of the incriminated method in unit tests.

**Comment by h3xstream** [comment]
Thanks for the notification this is really appreciate.

The good news is that it is already fixed. :smile:  I update to the latest API (SQ 5.6 LTS) few weeks ago and came across this deprecated API being removed.

**Comment by amandel** [comment]
Thanks for picking this up. 
Good to hear that this already is fixed, but I'm not sure if this is correct. 
This is about changes in the Java Analyzer with the upcoming to be released version 4.0 of the analyzer, not the SQ server itself.
I did my test with the version 3.3 of the findbugs plugin, which is the latest version available through the update center. I could not find a 3.4 for download.
